src/Aspire.Hosting/ProjectResourceBuilderExtensions.cs

```diff
@@ -491,4 +490,47 @@ private static void SetAspNetCoreUrls(this IResourceBuilder<ProjectResource> bui
             }
         });
     }
+
+    private static void SetBothPortsEnvVariables(this IResourceBuilder<ProjectResource> builder)
+    {
+        builder.WithEnvironment(context =>
+        {
+            builder.SetOnePortsEnvVariable(context, "HTTP_PORTS", "http");
+            builder.SetOnePortsEnvVariable(context, "HTTPS_PORTS", "https");
+        });
+    }
+
+    private static void SetOnePortsEnvVariable(this IResourceBuilder<ProjectResource> builder, EnvironmentCallbackContext context, string portEnvVariable, string scheme)
+    {
+        if (context.EnvironmentVariables.ContainsKey(portEnvVariable))
+        {
+            // If the user has already set that variable, we don't want to override it.
+            return;
+        }
+
+        var ports = new ReferenceExpressionBuilder();
+        var firstPort = true;
+
+        // Turn endpoint ports into a single environment variable
+        foreach (var e in builder.Resource.GetEndpoints().Where(e => IsValidAspNetCoreUrl(e.EndpointAnnotation)))
+        {
+            if (e.EndpointAnnotation.UriScheme == scheme && e.EndpointAnnotation.TargetPort != null)
```

Comment:
I don't think we need a port check at all.

Reply:
Interesting. So even if someone has `.WithHttpEndpoint(name: "hp1", port: 5001)` without a TargetPort, that should turn into an `HTTP_PORTS` entry. So I guess we're really collecting all named bindings in getting them in there? I don't have my head fully wrapped around this, but I'll take your word for it!

Reply:
Yep.